File: server/services/game-export/unreal/templates/source/world/WorldScaleManager.cpp
#include "WorldScaleManager.h"

void UWorldScaleManager::Initialize(FSubsystemCollectionBase& Collection)
{
    Super::Initialize(Collection);
    UE_LOG(LogTemp, Log, TEXT("[Insimul] WorldScaleManager initialized (terrain: %d)"), TerrainSize);
}

float UWorldScaleManager::GetSettlementRadius(int32 Population)
{
    struct FPopTier { int32 Min; int32 Max; float Radius; };
    static const FPopTier Tiers[] = {
        {0, 50, 20.f}, {51, 200, 35.f}, {201, 1000, 55.f},
        {1001, 5000, 80.f}, {5001, INT32_MAX, 120.f}
    };

    for (const auto& T : Tiers)
    {
        if (Population <= T.Max) return T.Radius;
    }
    return 20.f;
}

FString UWorldScaleManager::GetSettlementTier(int32 Population)
{
    if (Population < 100) return TEXT("hamlet");
    if (Population < 500) return TEXT("village");
    if (Population < 2000) return TEXT("town");
    if (Population < 10000) return TEXT("city");
    return TEXT("metropolis");
}

TArray<FVector> UWorldScaleManager::GenerateLotPositions(FVector SettlementPosition, float SettlementRadius, int32 LotCount, const TArray<FString>& StreetNames)
{
    TArray<FVector> Positions;
    if (LotCount <= 0) return Positions;

    const int32 Cols = FMath::CeilToInt(FMath::Sqrt(static_cast<float>(LotCount)));
    const int32 Rows = FMath::CeilToInt(static_cast<float>(LotCount) / Cols);
    const float LotSpacing = 20.f;
    const float GridWidth = (Cols - 1) * LotSpacing;
    const float GridHeight = (Rows - 1) * LotSpacing;

    FRandomStream Rand(GetTypeHash(FString::Printf(TEXT("%f_%f"), SettlementPosition.X, SettlementPosition.Z)));

    for (int32 i = 0; i < LotCount; i++)
    {
        const int32 Row = i / Cols;
        const int32 Col = i % Cols;

        float BaseX = SettlementPosition.X - GridWidth / 2.f + Col * LotSpacing;
        float BaseZ = SettlementPosition.Y - GridHeight / 2.f + Row * LotSpacing;

        float JitterX = (Rand.FRand() - 0.5f) * 4.f;
        float JitterZ = (Rand.FRand() - 0.5f) * 4.f;

        float LotX = BaseX + JitterX;
        float LotZ = BaseZ + JitterZ;

        // Push lots outside spawn clear radius
        float DX = LotX - SettlementPosition.X;
        float DZ = LotZ - SettlementPosition.Y;
        float Dist = FMath::Sqrt(DX * DX + DZ * DZ);
        if (Dist < SPAWN_CLEAR_RADIUS)
        {
            float Angle = Dist > 0.001f ? FMath::Atan2(DZ, DX) : (i * PI * 0.618f);
            LotX = SettlementPosition.X + FMath::Cos(Angle) * SPAWN_CLEAR_RADIUS;
            LotZ = SettlementPosition.Y + FMath::Sin(Angle) * SPAWN_CLEAR_RADIUS;
        }

        Positions.Add(FVector(LotX, LotZ, 0.f));
    }

    return Positions;
}

TArray<FVector> UWorldScaleManager::DistributeSettlements(FVector BoundsMin, FVector BoundsMax, FVector BoundsCenter, int32 SettlementCount, const TArray<float>& Radii, int32 WorldSeed)
{
    TArray<FVector> Positions;
    if (SettlementCount <= 0) return Positions;

    const float BoundsW = BoundsMax.X - BoundsMin.X;
    const float BoundsH = BoundsMax.Z - BoundsMin.Z;

    // Reserve 25% of the world radius as margin on each side so buildings
    // never approach the terrain edge (which shows as void/water on the minimap).
    const float Margin = FMath::Min(BoundsW, BoundsH) * 0.25f;
    const float SafeMinX = BoundsMin.X + Margin;
    const float SafeMaxX = BoundsMax.X - Margin;
    const float SafeMinZ = BoundsMin.Z + Margin;
    const float SafeMaxZ = BoundsMax.Z - Margin;

    FRandomStream Rand(WorldSeed);

    for (int32 Index = 0; Index < SettlementCount; Index++)
    {
        const float Radius = Radii.IsValidIndex(Index) ? Radii[Index] : 20.f;
        FVector Position;

        if (SettlementCount == 1)
        {
            // Single settlement: place exactly at world center
            Position = BoundsCenter;
        }
        else
        {
            int32 Attempts = 0;
            const int32 MaxAttempts = 50;
            bool bPlaced = false;

            while (Attempts < MaxAttempts)
            {
                const float X = SafeMinX + Rand.FRand() * FMath::Max(SafeMaxX - SafeMinX, 1.f);
                const float Z = SafeMinZ + Rand.FRand() * FMath::Max(SafeMaxZ - SafeMinZ, 1.f);
                Position = FVector(X, 0.f, Z);

                // Check if too close to other settlements
                bool bTooClose = false;
                for (int32 j = 0; j < Positions.Num(); j++)
                {
                    const float Dist = FVector::Dist(Position, Positions[j]);
                    const float OtherRadius = Radii.IsValidIndex(j) ? Radii[j] : 20.f;
                    if (Dist < (Radius + OtherRadius + 10.f))
                    {
                        bTooClose = true;
                        break;
                    }
                }

                if (!bTooClose)
                {
                    bPlaced = true;
                    break;
                }
                Attempts++;
            }

            // If couldn't find good position, use grid fallback centered in the safe zone
            if (!bPlaced)
            {
                const int32 Cols = FMath::CeilToInt(FMath::Sqrt(static_cast<float>(SettlementCount)));
                const int32 Row = Index / Cols;
                const int32 Col = Index % Cols;

                const float CellWidth = (SafeMaxX - SafeMinX) / Cols;
                const float CellHeight = (SafeMaxZ - SafeMinZ) / FMath::CeilToInt(static_cast<float>(SettlementCount) / Cols);

                Position = FVector(
                    SafeMinX + Col * CellWidth + CellWidth / 2.f,
                    0.f,
                    SafeMinZ + Row * CellHeight + CellHeight / 2.f
                );
            }
        }

        Positions.Add(Position);
    }

    return Positions;
}

void UWorldScaleManager::GenerateStreetAlignedSettlement(FVector SettlementPosition, float SettlementRadius, int32 LotCount, int32 BizCount, const TArray<FString>& StreetNames, const TArray<FVector>& ExistingStreetPoints)
{
    // TODO: Implement street-aligned placement (main street + side streets).
    // For now, falls back to grid+jitter via GenerateLotPositions.
    // ExistingStreetPoints can be used to align with a pre-existing street network.
    UE_LOG(LogTemp, Warning, TEXT("[Insimul] GenerateStreetAlignedSettlement not yet implemented in export template"));
}

int32 UWorldScaleManager::CalculateOptimalWorldSize(int32 CountryCount, int32 StateCount, int32 SettlementCount)
{
    // Minimum 1024 so that a single town's server-generated street grid
    // (mapSize 500-1000) fits comfortably within the world with margin.
    const float MaxEntities = FMath::Max3(
        static_cast<float>(CountryCount),
        static_cast<float>(StateCount) / 2.f,
        static_cast<float>(SettlementCount) / 5.f
    );

    if (MaxEntities <= 4.f) return 1024;
    if (MaxEntities <= 9.f) return 1536;
    if (MaxEntities <= 16.f) return 2048;
    if (MaxEntities <= 25.f) return 2560;
    return 3072;
}

File: server/services/game-export/unity/templates/scripts/world/WorldScaleManager.cs
using UnityEngine;
using Insimul.Data;

namespace Insimul.World
{
    public class WorldScaleManager : MonoBehaviour
    {
        [Header("World")]
        public int terrainSize = {{TERRAIN_SIZE}};
        public Color groundColor = new Color({{GROUND_COLOR_R}}f, {{GROUND_COLOR_G}}f, {{GROUND_COLOR_B}}f);
        public Color skyColor = new Color({{SKY_COLOR_R}}f, {{SKY_COLOR_G}}f, {{SKY_COLOR_B}}f);
        public Color roadColor = new Color({{ROAD_COLOR_R}}f, {{ROAD_COLOR_G}}f, {{ROAD_COLOR_B}}f);

        public void Initialize(InsimulWorldIR worldData)
        {
            terrainSize = worldData.geography.terrainSize;
            Debug.Log($"[Insimul] WorldScaleManager initialized (terrain: {terrainSize})");

            // Set skybox color
            RenderSettings.ambientSkyColor = skyColor;
            Camera.main.backgroundColor = skyColor;

            // Generate terrain from heightmap or flat fallback
            GenerateTerrain(worldData.geography.heightmap);
        }

        private void GenerateTerrain(float[][] heightmap)
        {
            var terrainGO = new GameObject("Terrain");
            terrainGO.transform.position = Vector3.zero;

            var meshGen = terrainGO.AddComponent<TerrainMeshGenerator>();
            meshGen.GenerateFromHeightmap(heightmap, terrainSize, groundColor);
        }

        public const float SPAWN_CLEAR_RADIUS = 15f;

        private static readonly (int min, int max, float radius)[] PopScale = new[]
        {
            (0, 50, 20f), (51, 200, 35f), (201, 1000, 55f),
            (1001, 5000, 80f), (5001, int.MaxValue, 120f)
        };

        public static float GetSettlementRadius(int population)
        {
            if (population <= 50) return 20f;
            if (population <= 200) return 35f;
            if (population <= 1000) return 55f;
            if (population <= 5000) return 80f;
            return 120f;
        }

        public static string GetSettlementTier(int population)
        {
            if (population < 100) return "hamlet";
            if (population < 500) return "village";
            if (population < 2000) return "town";
            if (population < 10000) return "city";
            return "metropolis";
        }

        public static Vector3[] GenerateLotPositions(Vector3 settlementPosition, float settlementRadius, int lotCount)
        {
            var positions = new Vector3[lotCount];
            if (lotCount <= 0) return positions;

            int cols = Mathf.CeilToInt(Mathf.Sqrt(lotCount));
            int rows = Mathf.CeilToInt((float)lotCount / cols);
            float lotSpacing = 20f;
            float gridWidth = (cols - 1) * lotSpacing;
            float gridHeight = (rows - 1) * lotSpacing;

            // Simple seeded random based on position
            System.Random rand = new System.Random(
                settlementPosition.GetHashCode());

            for (int i = 0; i < lotCount; i++)
            {
                int row = i / cols;
                int col = i % cols;

                float baseX = settlementPosition.x - gridWidth / 2f + col * lotSpacing;
                float baseZ = settlementPosition.z - gridHeight / 2f + row * lotSpacing;

                float jitterX = ((float)rand.NextDouble() - 0.5f) * 4f;
                float jitterZ = ((float)rand.NextDouble() - 0.5f) * 4f;

                float lotX = baseX + jitterX;
                float lotZ = baseZ + jitterZ;

                // Push lots outside spawn clear radius
                float dx = lotX - settlementPosition.x;
                float dz = lotZ - settlementPosition.z;
                float dist = Mathf.Sqrt(dx * dx + dz * dz);
                if (dist < SPAWN_CLEAR_RADIUS)
                {
                    float angle = dist > 0.001f
                        ? Mathf.Atan2(dz, dx)
                        : i * Mathf.PI * 0.618f;
                    lotX = settlementPosition.x + Mathf.Cos(angle) * SPAWN_CLEAR_RADIUS;
                    lotZ = settlementPosition.z + Mathf.Sin(angle) * SPAWN_CLEAR_RADIUS;
                }

                positions[i] = new Vector3(lotX, 0f, lotZ);
            }

            return positions;
        }

        /// <summary>
        /// Distribute settlements within territory bounds, using 25% margin and
        /// center-placement for single settlements.
        /// </summary>
        public static Vector3[] DistributeSettlements(
            Vector3 boundsMin, Vector3 boundsMax, Vector3 boundsCenter,
            int settlementCount, float[] radii, int worldSeed)
        {
            var positions = new Vector3[settlementCount];
            if (settlementCount <= 0) return positions;

            float boundsW = boundsMax.x - boundsMin.x;
            float boundsH = boundsMax.z - boundsMin.z;

            // Reserve 25% of the world radius as margin on each side so buildings
            // never approach the terrain edge (which shows as void/water on the minimap).
            float margin = Mathf.Min(boundsW, boundsH) * 0.25f;
            float safeMinX = boundsMin.x + margin;
            float safeMaxX = boundsMax.x - margin;
            float safeMinZ = boundsMin.z + margin;
            float safeMaxZ = boundsMax.z - margin;

            System.Random rand = new System.Random(worldSeed);

            for (int index = 0; index < settlementCount; index++)
            {
                float radius = (index < radii.Length) ? radii[index] : 20f;
                Vector3 position;

                if (settlementCount == 1)
                {
                    // Single settlement: place exactly at world center
                    position = boundsCenter;
                }
                else
                {
                    int attempts = 0;
                    const int maxAttempts = 50;
                    bool placed = false;
                    position = Vector3.zero;

                    while (attempts < maxAttempts)
                    {
                        float x = safeMinX + (float)rand.NextDouble() * Mathf.Max(safeMaxX - safeMinX, 1f);
                        float z = safeMinZ + (float)rand.NextDouble() * Mathf.Max(safeMaxZ - safeMinZ, 1f);
                        position = new Vector3(x, 0f, z);

                        // Check if too close to other settlements
                        bool tooClose = false;
                        for (int j = 0; j < index; j++)
                        {
                            float dist = Vector3.Distance(position, positions[j]);
                            float otherRadius = (j < radii.Length) ? radii[j] : 20f;
                            if (dist < (radius + otherRadius + 10f))
                            {
                                tooClose = true;
                                break;
                            }
                        }

                        if (!tooClose)
                        {
                            placed = true;
                            break;
                        }
                        attempts++;
                    }

                    // If couldn't find good position, use grid fallback centered in the safe zone
                    if (!placed)
                    {
                        int cols = Mathf.CeilToInt(Mathf.Sqrt(settlementCount));
                        int row = index / cols;
                        int col = index % cols;

                        float cellWidth = (safeMaxX - safeMinX) / cols;
                        float cellHeight = (safeMaxZ - safeMinZ) / Mathf.CeilToInt((float)settlementCount / cols);

                        position = new Vector3(
                            safeMinX + col * cellWidth + cellWidth / 2f,
                            0f,
                            safeMinZ + row * cellHeight + cellHeight / 2f
                        );
                    }
                }

                positions[index] = position;
            }

            return positions;
        }

        /// <summary>
        /// Calculate recommended world size based on entity counts.
        /// Minimum 1024 so that a single town's server-generated street grid
        /// (mapSize 500-1000) fits comfortably within the world with margin.
        /// </summary>
        public static int CalculateOptimalWorldSize(int countryCount, int stateCount, int settlementCount)
        {
            float maxEntities = Mathf.Max(
                countryCount,
                Mathf.Max(stateCount / 2f, settlementCount / 5f));

            if (maxEntities <= 4f) return 1024;
            if (maxEntities <= 9f) return 1536;
            if (maxEntities <= 16f) return 2048;
            if (maxEntities <= 25f) return 2560;
            return 3072;
        }
    }
}
